### PR TITLE
Add rules_python 0.4.0

### PR DESCRIPTION
**Pull request by meteorcloudy**
Working towards: https://github.com/bazelbuild/bazel-central-registry/issues/7

### CODE REVIEW

File: modules/rules_python/metadata.json
Comment (alexeagle) [inline modules/rules_python/metadata.json:5]: this is wrong, the bazel team has not been participating **at all** in rules_python for the last year.

Maintainers are me, @hrfuller and @thundergolfer

How will you prevent this kind of mistake for future PRs to this repo? I've already tracked down the maintainers of all the rules as part of the rules authors SIG so I think we should work from that.

Comment (meteorcloudy) [reply modules/rules_python/metadata.json:5]: Sorry, this is auto-generated by a script I wrote: https://github.com/bazelbuild/bazel-central-registry/blob/main/tools/add_module.py#L136 
I was actually thinking about asking if you want to maintain this. 
I guess fixing this script is the first step to avoid this kind of error...

Comment (Wyverald) [reply modules/rules_python/metadata.json:5]: 1. "Maintainer" here does *not* mean a contributor to the project itself, rather a person who looks after the *module definition* here in the registry. When a new version is submitted, or new patches applied, the maintainers should be notified and approve the request. That much can definitely be done by the Bazel team. With that said, if the original contributors of the project are actively interested in maintaining the Bazel modules (very likely for rulesets, very unlikely for C++ projects like zlib), they can definitely be added to the maintainer list.
2. The rule author SIG is definitely a good starting point to assign maintainers with active interest (though as noted elsewhere, the registry isn't for rulesets only).
3. The central registry is still in alpha phase. We haven't opened this up yet; this PR is just part of the process to check in some basic Bazel modules so that we can build Bazel itself with the new system. We fully intend to assign the maintainers to interested parties before the registry is opened up for public use.

Comment (alexeagle) [reply modules/rules_python/metadata.json:5]: I think the concept you're looking for is who approves PRs against this file, and that belongs in a CODEOWNERS file. It's super confusing to put that metadata here, I think most ppl would be like me and assume that modules/rules_python/metadata.json is the metadata for the rules_python module

Comment (meteorcloudy) [reply modules/rules_python/metadata.json:5]: Interesting, I didn't expect it to be so confusing.

Similar to https://packages.debian.org/sid/libgrpc++-dev, when you see the maintainer in this Debian package, do you think they must be the author of the project?

Maybe this can be clarified in documentation?

**Comment by meteorcloudy** [comment]
Updates:

1) I'm leaving the maintainer for this module empty for now.

2) I added a module extension through a patch to propagate dependencies introduced by "pip_install_dependencies" so that I can build some target. This is still highly experimental, but it's a good proof that our module extension mechanism is a good bridge between repo rules and Bzlmod.

File: modules/rules_python/0.4.0/presubmit.yml
Comment (alexeagle) [inline modules/rules_python/0.4.0/presubmit.yml:2]: rules_python already integration tests the release artifact it pushes to github, using bazel_integration_test within the repo.

would we want to repeat that testing here in the registry? I would expect that registry testing for releases is a poor substitute for the rules testing that themselves, since you don't notice a problem until you send a PR to this repo (which is probably after the release was already published)

File: modules/rules_python/metadata.json
Comment (alexeagle) [reply modules/rules_python/metadata.json]: Yeah I understand what you're getting at now. It's possible for someone other than the rules authors to own the registration of those rules. Do you think that would be common though? Won't rules authors have to come make a PR to this repo anyway every time they cut a release?

Comment (meteorcloudy) [reply modules/rules_python/metadata.json]: https://github.com/bazelbuild/bazel-central-registry/tree/maintainer#module-maintainer

> Do you think that would be common though? 

It's totally possible for some non-Bazel projects. For example, many of TensorFlow's dependencies: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/workspace2.bzl#L134

> Won't rules authors have to come make a PR to this repo anyway every time they cut a release?

That's right. For rules authors, they should definitely be the maintainer for their entries.

File: modules/rules_python/0.4.0/presubmit.yml
Comment (meteorcloudy) [reply modules/rules_python/0.4.0/presubmit.yml:2]: I had some explanation for this file here: https://github.com/bazelbuild/bazel-central-registry/tree/maintainer#presubmityml

> rules_python already integration tests the release artifact it pushes to github, using bazel_integration_test within the repo.

That is good to known as I just noticed you also strip out some sources in the release archive.

> would we want to repeat that testing here in the registry?

In rules_python's case, I think it's safe to exclude most of your test cases (they are not available in the release archive anyway). But it's still good to specify some build targets that you want to expose to users here, just as a sanity check and make sure nothing serious was going wrong. 

> I would expect that registry testing for releases is a poor substitute for the rules testing that themselves

Again, we don't intend this to be a substitution for your own testing.

The presubmit.yml probably makes more sense for you in the following case:

Imagine there is a project X, which depends on rules_python. X has a presubmit.yml to specify some heavily used build & test targets. When rules_python has a new release, we can upgrade rules_python version for X to see if those targets still pass. If not, either we discovered a bug in rules_python or we can notify the downstream user to migrate for the incompatible change. This can even be done before you release the new version by creating a new branch of this registry.

File: modules/rules_python/0.4.0/patches/propagate_pip_install_dependencies.patch
Comment (Wyverald) [inline modules/rules_python/0.4.0/patches/propagate_pip_install_dependencies.patch:49]: we actually don't need no-op tags here if all we want to do is to always install those dependencies. So you can actually get rid of this tag class (and line 53 altogether)

Comment (meteorcloudy) [reply modules/rules_python/0.4.0/patches/propagate_pip_install_dependencies.patch:49]: Oh, do you mean as long as `pip_install = use_extension("@rules_python//bzlmod:extensions.bzl", "pip_install")` is called, the implementation function will be executed?

Comment (Wyverald) [reply modules/rules_python/0.4.0/patches/propagate_pip_install_dependencies.patch:49]: yes, using an extension is enough to trigger its impl function.

Comment (meteorcloudy) [reply modules/rules_python/0.4.0/patches/propagate_pip_install_dependencies.patch:49]: Thanks, done!

**Review by Wyverald — APPROVED** [review]

**Comment by meteorcloudy** [comment]
@alexeagle I'm happy to discuss further if you have any more concerns. But I'm merging this to unblock checking in other modules that depends on rules_python.

File: modules/rules_python/0.4.0/presubmit.yml
Comment (alexeagle) [reply modules/rules_python/0.4.0/presubmit.yml:2]: okay, thanks for explaining. My only remaining concern is who is going to fix that patch file when you get inevitable conflicts from upstream. Or do you plan to send a PR to rules_python?

Comment (meteorcloudy) [reply modules/rules_python/0.4.0/presubmit.yml:2]: Oh, yes, this will be upstreamed to rules_python definitely. It's just I'm patching 0.4.0 so I'm storing it in the registry for now.

My plan is to write some migration guide/tool to Bzlmod first, then the community can help migrating together, of course we'll be helping as well by sending PRs.

**Pull request merged**
